packages/next/src/shared/lib/router/action-queue.ts

```diff
@@ -0,0 +1,195 @@
+
+function runRemainingActions(
+  actionQueue: AppRouterActionQueue,
+  setState: DispatchStatePromise
+) {
+  if (actionQueue.pending !== null) {
+    actionQueue.pending = actionQueue.pending.next
+    if (actionQueue.pending !== null) {
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      runAction({
+        actionQueue,
+        action: actionQueue.pending,
+        setState,
+      })
+    }
+  }
+}
+
+async function runAction({
+  actionQueue,
+  action,
+  setState,
+}: {
+  actionQueue: AppRouterActionQueue
+  action: ActionQueueNode
+  setState: DispatchStatePromise
+}) {
+  const prevState = actionQueue.state
+  if (!prevState) {
+    // This shouldn't happen as the state is initialized in the dispatcher if it's not set
+    throw new Error('Invariant: Router state not initialized')
+  }
+
+  actionQueue.pending = action
+  actionQueue.last = action
+
+  const payload = action.payload
+  const actionResult = actionQueue.action(prevState, payload)
+
+  function handleResult(nextState: AppRouterState) {
+    // if we discarded this action, the state should also be discarded
+    if (action.discarded) {
+      if (actionQueue.needsRefresh) {
+        actionQueue.needsRefresh = false
+        actionQueue.dispatch(
+          {
+            type: ACTION_REFRESH,
+            cache: createEmptyCacheNode(),
+            mutable: {},
+            origin: window.location.origin,
+          },
+          setState
+        )
+      }
+      return
+    }
+
+    actionQueue.state = nextState
+
+    if (actionQueue.devToolsInstance) {
+      actionQueue.devToolsInstance.send(payload, nextState)
+    }
+
+    runRemainingActions(actionQueue, setState)
+    action.resolve(nextState)
+  }
+
+  // if the action is a promise, set up a callback to resolve it
+  if (isThenable(actionResult)) {
+    actionResult.then(handleResult, (err) => {
+      runRemainingActions(actionQueue, setState)
+      action.reject(err)
+    })
+  } else {
+    handleResult(actionResult)
+  }
+}
+
+function dispatchAction(
+  actionQueue: AppRouterActionQueue,
+  payload: ReducerActions,
+  setState: DispatchStatePromise
+) {
+  let resolvers: {
+    resolve: (value: AppRouterState | PromiseLike<AppRouterState>) => void
+    reject: (reason: any) => void
+  }
+
+  // Create the promise and assign the resolvers to the object.
+  const deferredPromise = new Promise<AppRouterState>((resolve, reject) => {
+    resolvers = { resolve, reject }
+  })
+
+  const newAction: ActionQueueNode = {
+    payload,
+    next: null,
+    resolve: resolvers!.resolve,
+    reject: resolvers!.reject,
+  }
+
+  startTransition(() => {
+    // we immediately notify React of the pending promise -- the resolver is attached to the action node
+    // and will be called when the associated action promise resolves
+    setState(deferredPromise)
+  })
+
+  if (payload.type === ACTION_NAVIGATE && actionQueue.pending !== null) {
```

Comment:
Factoring nit: You can combine this with the if/else branches below. That way you only check once if the queue is empty (right now you're checking twice):

```js
// Check if the queue is empty
  if (actionQueue.pending === null) {
    // The queue is empty, so add the action and start it immediately
    runAction({
      actionQueue,
      action: newAction,
      setState,
    })
  } else if (payload.type === ACTION_NAVIGATE) {
    // Navigations take priority over any pending actions.
    // Mark the pending action as discarded (so the state is never applied) and start the navigation action immediately.
    actionQueue.pending.discarded = true

    // if the pending action was a server action, mark the queue as needing a refresh once events are processed
    actionQueue.needsRefresh =
      actionQueue.pending.payload.type === ACTION_SERVER_ACTION

    runAction({
      actionQueue,
      action: newAction,
      setState,
    })
  } else {
    // The queue is not empty, so add the action to the end of the queue
    // It will be started by finishRunningAction after the previous action finishes
    if (actionQueue.last !== null) {
      actionQueue.last.next = newAction
    }
    actionQueue.last = newAction
  }
```